server/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java

```diff
@@ -277,4 +120,67 @@ public void onFailure(Exception e) {
         }
     }
 
+    /**
+     * A task to be executed, along with callbacks for the executor to record the outcome of this task's execution. The executor must
+     * call exactly one of these methods for every task in its batch.
+     */
+    interface TaskContext<T extends ClusterStateTaskListener> {
+
+        /**
+         * @return the task to be executed.
+         */
+        T getTask();
+
+        /**
+         * Record that the task succeeded.
+         * <p>
+         * Note that some tasks implement {@link ClusterStateAckListener} and can listen for acks themselves. If so, you may not use this
+         * method and must instead call {@link #success(ActionListener, ClusterStateAckListener)}, passing the task itself as the {@code
+         * clusterStateAckListener} argument.
+         *
+         * @param publishListener A listener for the completion of the resulting cluster state publication. This listener is completed
+         *                        with the cluster state that was published (or the publication exception that occurred) in the thread
+         *                        context in which the task was submitted. The task's {@link
+         *                        ClusterStateTaskListener#clusterStateProcessed} method is not called directly by the master service,
+         *                        nor is {@link ClusterStateTaskListener#onFailure} once the task execution has succeeded, but legacy
+         *                        implementations may supply a listener which calls those methods.
+         *                        <p>
+         *                        The listener should prefer not to use the published state for things like determining the result of a
+         *                        task. The task may have been executed as part of a batch, and later tasks in the batch may overwrite
+         *                        the results from earlier tasks. Instead the listener should independently capture the information it
+         *                        needs to properly process the completion of a cluster state update.
+         */
+        // TODO remove all remaining usages of the published state and then make publishListener an ActionListener<Void>
```

Comment:
It would be great to add a link to the issue (if there is one) to be able to track when this TODO could be removed

Reply:
:+1: done in 112bf681642